test/utils/ruleset-test-harness.ts

```diff
@@ -0,0 +1,151 @@
+  expectFS, ExpectObserver, setupGlobals, expectGlobal,
+} from './mocha';
+
+export interface RulesetTestHarness {
+  setUp: (rules: any, dataLayer: any) => Promise<void>;
+  tearDown: () => Promise<void>;
+  execute: (action: (args: any[]) => void, args?: any[]) => Promise<void>;
+  popEvent: (timeoutMs?: number) => Promise<any[]>;
+}
+
+interface RulesetTestEnvironment {
+  name: string;
+  createTestHarness: (rules: any, dataLayer: any) => Promise<RulesetTestHarness>;
+  tearDown: () => Promise<void>;
+}
+
+const isBrowserTest = process.env.DLO_RUN_BROWSER_TESTS;
+const dloScriptSrc = process.env.PLAYWRIGHT_DLO_SCRIPT_SRC;
+
+const nodeTestHarness: RulesetTestHarness = {
+  setUp: (rules: any, dataLayer: any) => {
+    setupGlobals(
+      Object.keys(dataLayer).map((key) => [key, dataLayer[key]]),
+    );
+    ExpectObserver.getInstance().create({ rules });
+    return Promise.resolve();
+  },
+
+  tearDown: () => {
+    ExpectObserver.getInstance().cleanup();
+    return Promise.resolve();
+  },
+
+  execute: (action: (args: any[]) => void, args?: any[]) => {
+    action(args || []);
+    return Promise.resolve();
+  },
+
+  popEvent: async (timeoutMs: number = 1000) => new Promise<any[]>((resolve, reject) => {
+    const startTime = new Date().getTime();
+
+    const checkForEvents = setInterval(() => {
+      const currentTime = new Date().getTime();
+      if (currentTime - startTime >= timeoutMs) {
+        clearInterval(checkForEvents);
+        reject(new Error(`Timeout ${timeoutMs} exceeded waiting for FS.event call.`));
+      }
+
+      if (expectGlobal('FS').callQueues.event.length) {
+        clearInterval(checkForEvents);
+        resolve(expectFS('event'));
+      }
+    }, 50);
+  }),
+};
+
+class BrowserTestHarness implements RulesetTestHarness {
+  private browser: Browser = null!;
+
+  private page: Page = null!;
+
+  constructor(private readonly browserPromise: Promise<Browser>) {
+    if (!dloScriptSrc) {
+      throw new Error(
+        'PLAYWRIGHT_DLO_SCRIPT_SRC must be configured with a DLO library script location for browser tests.',
+      );
+    }
+  }
+
+  async setUp(rules: any, dataLayer: any) {
+    this.browser = await this.browserPromise;
+    this.page = await this.browser.newPage();
+
+    await this.page.evaluate(([localRules, localDataLayer, localDloScriptSrc]) => {
+      // This allows node and playwright tests to reference the same global/window name.
+      const globalThis: any = window;
+
+      globalThis._dlo_rules = localRules;
+
+      Object.keys(localDataLayer).forEach((key) => {
+        globalThis[key] = localDataLayer[key];
+      });
+
+      globalThis.events = [];
+
+      globalThis.FS = {
+        event: (...args: any) => {
+          globalThis.events.push(args);
+        },
+      };
+
+      const dloScriptTag = document.createElement('script');
+      dloScriptTag.src = localDloScriptSrc;
+      document.body.appendChild(dloScriptTag);
+    }, [rules, dataLayer, dloScriptSrc]);
```

Comment:
why dont we want the browser scoped rules/dataLayer/dloScriptSrc? I might be understanding this incorrectly but don't the node tests use the browser scoped rules/dataLayer/dloScriptSrc?

Reply:
We send the browser our own local rules; however, the DLO library is loaded from the given JS file (`dloScriptSrc`) for browser tests. Here's how I think about the testing breakdown relative to the DLO dev lifecycle:

- When developing DLO locally, you'll typically run `npm test`. This executes test logic without running it through a browser, using local rulesets and local library logic (i.e. the working directory). This is a faster feedback loop and is also the test command run via CI.
- When testing DLO release candidates (e.g. a candidate built locally or deployed to a staging environment), you'll typically run `PLAYWRIGHT_DLO_SCRIPT_SRC=http... npm run test:browser`. This executes the same test logic through target headless browsers, using the same local rulesets but pulling in the DLO library logic from the given JS file (aka release candidate). The goal here is to build confidence in a transpiled JS release candidate rather than the local working directory/TS files.

Let me know what you think.

Reply:
That makes sense, thank you for clarifying.